kernel/src/fs/notify/mod.rs

```diff
@@ -0,0 +1,363 @@
+    /// List of fsnotify subscribers.
+    subscribers: RwLock<Vec<Arc<dyn FsnotifySubscriber>>>,
+    /// All interested fsnotify events (aggregated from all subscribers).
+    /// Stored as AtomicU32 for lock-free reads, but accessed as FsnotifyEvents.
+    all_interested_events: AtomicU32,
+}
+
+impl Debug for FsnotifyPublisher {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        let subscribers = self.subscribers.read();
+        write!(
+            f,
+            "FsnotifyPublisher: num_subscribers: {}",
+            subscribers.len()
+        )?;
+        Ok(())
+    }
+}
+
+impl Default for FsnotifyPublisher {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl FsnotifyPublisher {
+    pub fn new() -> Self {
+        Self {
+            subscribers: RwLock::new(Vec::new()),
+            all_interested_events: AtomicU32::new(0),
+        }
+    }
+
+    /// Adds a subscriber to this publisher.
+    pub fn add_subscriber(&self, subscriber: Arc<dyn FsnotifySubscriber>) {
+        let mut subscribers = self.subscribers.write();
+        // Update aggregated events with new subscriber's interested events
+        self.all_interested_events
+            .fetch_or(subscriber.interested_events().bits(), Ordering::Relaxed);
+        subscribers.push(subscriber);
+    }
+
+    /// Removes a subscriber from this publisher.
+    pub fn remove_subscriber(&self, subscriber: &Arc<dyn FsnotifySubscriber>) {
+        let mut subscribers = self.subscribers.write();
+        subscribers.retain(|m| !Arc::ptr_eq(m, subscriber));
+        subscriber.deliver_event(FsnotifyEvents::IN_IGNORED, None);
+
+        // Recalculate aggregated events from remaining subscribers
+        self.recalc_interested_events(&subscribers);
+    }
+
+    /// Removes all subscribers from this publisher.
+    pub fn remove_all_subscribers(&self) -> usize {
+        let mut subscribers = self.subscribers.write();
+        // Notify all subscribers about IN_IGNORED event before removing them.
+        for subscriber in subscribers.iter() {
+            subscriber.deliver_event(FsnotifyEvents::IN_IGNORED, None);
+        }
+        // Return the number of subscribers removed.
+        let num_subscribers = subscribers.len();
+        // Clear all subscribers.
+        subscribers.clear();
+
+        // Clear aggregated events since there are no more subscribers
+        self.all_interested_events.store(0, Ordering::Relaxed);
+
+        num_subscribers
+    }
+
+    /// Broadcasts an event to all the subscribers of this publisher.
+    pub fn publish_event(&self, events: FsnotifyEvents, name: Option<String>) {
+        // Fast path: check aggregated events first
+        let interested =
+            FsnotifyEvents::from_bits_truncate(self.all_interested_events.load(Ordering::Relaxed));
+        // If the aggregated events does not intersect with the events, return early.
+        if !interested.intersects(events) {
```

Comment:
Another example of worthless comments. The comment is even longer but less straightforward than the code itself...